NoClassDefFoundError on generation of Eclipse formatter settings from checkstyle.xml

Eclipse 2018-09 (4.9.0) Java EE Edition
CheckStyle 8.12 plugin

Trying to generate my Eclipse formatter settings from my checkstyle.xml is generating this error.

```
!ENTRY org.eclipse.core.jobs 4 2 2018-10-22 08:39:30.452
!MESSAGE An internal error occurred during: "transformCheckstyle".
!STACK 0
java.lang.NoClassDefFoundError: org/eclipse/jdt/internal/ui/preferences/PreferencesAccess
	at net.sf.eclipsecs.core.transformer.FormatterConfigWriter.writeCleanupSettings(FormatterConfigWriter.java:95)
	at net.sf.eclipsecs.core.transformer.FormatterConfigWriter.writeSettings(FormatterConfigWriter.java:89)
	at net.sf.eclipsecs.core.transformer.FormatterConfigWriter.<init>(FormatterConfigWriter.java:81)
	at net.sf.eclipsecs.core.transformer.CheckstyleTransformer.transformRules(CheckstyleTransformer.java:124)
	at net.sf.eclipsecs.core.jobs.TransformCheckstyleRulesJob.runInWorkspace(TransformCheckstyleRulesJob.java:117)
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:42)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

Just FYI I've also just hit this one (currently running under Java 9):

```
eclipse.buildId=4.9.0.I20180906-0745
java.version=9.0.4
java.vendor=Oracle Corporation
BootLoader constants: OS=linux, ARCH=x86_64, WS=gtk, NL=en_IN
Framework arguments:  -product org.eclipse.epp.package.java.product
Command-line arguments:  -os linux -ws gtk -arch x86_64 -product org.eclipse.epp.package.java.product

org.eclipse.core.jobs
Error
Fri Nov 23 14:54:43 CET 2018
An internal error occurred during: "transformCheckstyle".

java.lang.NoClassDefFoundError: org/eclipse/jdt/internal/ui/preferences/PreferencesAccess
	at net.sf.eclipsecs.core.transformer.FormatterConfigWriter.writeCleanupSettings(FormatterConfigWriter.java:95)
	at net.sf.eclipsecs.core.transformer.FormatterConfigWriter.writeSettings(FormatterConfigWriter.java:89)
	at net.sf.eclipsecs.core.transformer.FormatterConfigWriter.<init>(FormatterConfigWriter.java:81)
	at net.sf.eclipsecs.core.transformer.CheckstyleTransformer.transformRules(CheckstyleTransformer.java:124)
	at net.sf.eclipsecs.core.jobs.TransformCheckstyleRulesJob.runInWorkspace(TransformCheckstyleRulesJob.java:117)
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:42)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.lang.ClassNotFoundException: org.eclipse.jdt.internal.ui.preferences.PreferencesAccess cannot be found by net.sf.eclipsecs.core_8.12.0.201808161509
	at org.eclipse.osgi.internal.loader.BundleLoader.findClassInternal(BundleLoader.java:460)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:422)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:414)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.loadClass(ModuleClassLoader.java:153)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:496)
	... 7 more
```

--- a/net.sf.eclipsecs.core/src/net/sf/eclipsecs/core/transformer/FormatterConfigWriter.java
+++ b/net.sf.eclipsecs.core/src/net/sf/eclipsecs/core/transformer/FormatterConfigWriter.java
@@ -20,48 +20,32 @@
 
 package net.sf.eclipsecs.core.transformer;
 
-import java.util.ArrayList;
-import java.util.List;
+import java.io.InputStream;
 import java.util.Map;
+
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.TransformerException;
 
 import net.sf.eclipsecs.core.util.CheckstyleLog;
 
+import org.eclipse.core.resources.IFile;
 import org.eclipse.core.resources.IProject;
 import org.eclipse.core.runtime.CoreException;
-import org.eclipse.core.runtime.preferences.IScopeContext;
-import org.eclipse.jdt.core.JavaCore;
-import org.eclipse.jdt.internal.corext.fix.CleanUpConstants;
-import org.eclipse.jdt.internal.corext.fix.CleanUpPreferenceUtil;
-import org.eclipse.jdt.internal.ui.preferences.PreferencesAccess;
-import org.eclipse.jdt.internal.ui.preferences.cleanup.CleanUpProfileManager;
-import org.eclipse.jdt.internal.ui.preferences.cleanup.CleanUpProfileVersioner;
-import org.eclipse.jdt.internal.ui.preferences.formatter.FormatterProfileManager;
-import org.eclipse.jdt.internal.ui.preferences.formatter.FormatterProfileStore;
-import org.eclipse.jdt.internal.ui.preferences.formatter.IProfileVersioner;
-import org.eclipse.jdt.internal.ui.preferences.formatter.ProfileManager;
-import org.eclipse.jdt.internal.ui.preferences.formatter.ProfileManager.CustomProfile;
-import org.eclipse.jdt.internal.ui.preferences.formatter.ProfileManager.Profile;
-import org.eclipse.jdt.internal.ui.preferences.formatter.ProfileStore;
-import org.eclipse.jdt.internal.ui.preferences.formatter.ProfileVersioner;
-import org.osgi.service.prefs.BackingStoreException;
 
 /**
- * Class for writing a new eclipse-configuration-file. Gets used by class Transformer. A new
- * eclipse-formatter-profile gets added.
+ * Class for writing a new eclipse-configuration-file. Gets used by class Transformer. Two
+ * eclipse-formatter-profile files gets written to the project root.
  *
+ * @author Alexandros Karypidis
  * @author Lukas Frena
  * @author Lars Ködderitzsch
  */
-@SuppressWarnings("restriction")
 public class FormatterConfigWriter {
 
-  private static final String JDT_UI_PLUGINID = "org.eclipse.jdt.ui";
-
+  /** Constant for show generated code. */
+  private static final String CS_GENERATED = "CheckStyle-Generated ";
   /** A eclipse-configuration. */
   private final FormatterConfiguration mConfiguration;
-
-  /** Name of new createt profile. */
-  private final String mNewProfileName;
 
   private IProject mProject;
 
@@ -77,135 +61,58 @@ public class FormatterConfigWriter {
     mConfiguration = settings;
     mProject = project;
 
-    mNewProfileName = "eclipse-cs " + mProject.getName();
     writeSettings();
   }
 
   /**
-   * Method for writing all settings to disc. Also activates new profile.
+   * Method for persisting all settings to files.
    */
   private void writeSettings() {
-    // read the Eclipse-Preferences for manipulation
     writeCleanupSettings(mConfiguration.getCleanupSettings());
     writeFormatterSettings(mConfiguration.getFormatterSettings());
   }
 
+  /**
+   * Method for writing all cleanup settings to disc.
+   *
+   * @param settings
+   *          All the settings.
+   */
   private void writeCleanupSettings(final Map<String, String> settings) {
-
-    PreferencesAccess access = PreferencesAccess.getOriginalPreferences();
-
-    IScopeContext instanceScope = access.getInstanceScope();
-    IScopeContext scope = access.getProjectScope(mProject);
-
-    IProfileVersioner versioner = new CleanUpProfileVersioner();
-    ProfileStore profilesStore = new ProfileStore(CleanUpConstants.CLEANUP_PROFILES, versioner);
+    final IFile settingsFile = mProject.getFile(mProject.getName() + "-cs-cleanup.xml");
     try {
-
-      List<Profile> profiles = profilesStore.readProfiles(instanceScope);
-
-      if (profiles == null) {
-        profiles = new ArrayList<>();
-      }
-      profiles.addAll(CleanUpPreferenceUtil.getBuiltInProfiles());
-
-      ProfileManager manager = new CleanUpProfileManager(profiles, scope, access, versioner);
-
-      CustomProfile myProfile = (CustomProfile) manager
-              .getProfile(ProfileManager.ID_PREFIX + mNewProfileName);
-
-      if (myProfile == null) {
-        // take current settings and create new profile
-        Profile current = manager.getSelected();
-        myProfile = new CustomProfile(mNewProfileName, current.getSettings(),
-                versioner.getCurrentVersion(), versioner.getProfileKind());
-        manager.addProfile(myProfile);
-      }
-
-      Map<String, String> joinedSettings = myProfile.getSettings();
-      joinedSettings.putAll(settings);
-
-      myProfile.setSettings(joinedSettings);
-      manager.setSelected(myProfile);
-
-      // writes profiles to the workspace profile store
-      profilesStore.writeProfiles(manager.getSortedProfiles(), instanceScope);
-
-      // commits changes to the project profile settings
-      manager.commitChanges(scope);
-
-      scope.getNode(JDT_UI_PLUGINID).flush();
-      scope.getNode(JavaCore.PLUGIN_ID).flush();
-      if (scope != instanceScope) {
-        instanceScope.getNode(JDT_UI_PLUGINID).flush();
-        instanceScope.getNode(JavaCore.PLUGIN_ID).flush();
-      }
-
-    } catch (CoreException e) {
-      CheckstyleLog.log(e, "Error storing cleanup profile");
-    } catch (BackingStoreException e) {
-      CheckstyleLog.log(e, "Error storing cleanup profile");
+      final InputStream stream = XmlProfileWriter.writeCleanupProfileToStream(
+              CS_GENERATED + mProject.getName(), settings);
+      createOrUpdateFile(settingsFile, stream);
+    } catch (CoreException | TransformerException | ParserConfigurationException exc) {
+      CheckstyleLog.log(exc, "Error saving cleanup profile");
     }
   }
 
   /**
-   * Method for writing all formatter-settings to disc.
+   * Method for writing all formatter settings to disc.
    *
    * @param settings
    *          All the settings.
    */
   private void writeFormatterSettings(final Map<String, String> settings) {
-
-    PreferencesAccess access = PreferencesAccess.getOriginalPreferences();
-
-    IScopeContext instanceScope = access.getInstanceScope();
-    IScopeContext scope = access.getProjectScope(mProject);
-
-    IProfileVersioner versioner = new ProfileVersioner();
-    ProfileStore profilesStore = new FormatterProfileStore(versioner);
+    final IFile settingsFile = mProject.getFile(mProject.getName() + "-cs-formatter.xml");
     try {
-
-      List<Profile> profiles = profilesStore.readProfiles(instanceScope);
-
-      if (profiles == null) {
-        profiles = new ArrayList<>();
-      }
-
-      ProfileManager manager = new FormatterProfileManager(profiles, scope, access, versioner);
-
-      CustomProfile myProfile = (CustomProfile) manager
-              .getProfile(ProfileManager.ID_PREFIX + mNewProfileName);
-
-      if (myProfile == null) {
-        // take current settings and create new profile
-        Profile current = manager.getSelected();
-        myProfile = new CustomProfile(mNewProfileName, current.getSettings(),
-                versioner.getCurrentVersion(), versioner.getProfileKind());
-        manager.addProfile(myProfile);
-      }
-
-      Map<String, String> joinedSettings = myProfile.getSettings();
-      joinedSettings.putAll(settings);
-
-      myProfile.setSettings(joinedSettings);
-      manager.setSelected(myProfile);
-
-      // writes profiles to the workspace profile store
-      profilesStore.writeProfiles(manager.getSortedProfiles(), instanceScope);
-
-      // commits changes to the project profile settings
-      manager.commitChanges(scope);
-
-      scope.getNode(JDT_UI_PLUGINID).flush();
-      scope.getNode(JavaCore.PLUGIN_ID).flush();
-      if (scope != instanceScope) {
-        instanceScope.getNode(JDT_UI_PLUGINID).flush();
-        instanceScope.getNode(JavaCore.PLUGIN_ID).flush();
-      }
-
-    } catch (CoreException e) {
-      CheckstyleLog.log(e, "Error storing formatter profile");
-    } catch (BackingStoreException e) {
-      CheckstyleLog.log(e, "Error storing formatter profile");
+      final InputStream stream = XmlProfileWriter.writeFormatterProfileToStream(
+              CS_GENERATED + mProject.getName(), settings);
+      createOrUpdateFile(settingsFile, stream);
+    } catch (CoreException | TransformerException | ParserConfigurationException exc) {
+      CheckstyleLog.log(exc, "Error saving formatter profile");
     }
   }
+
+  private static void createOrUpdateFile(IFile settingsFile, InputStream stream)
+          throws CoreException {
+    if (settingsFile.exists()) {
+      settingsFile.setContents(stream, true, false, null);
+    } else {
+      settingsFile.create(stream, true, null);
+    }
+  }
+
 }

--- a/net.sf.eclipsecs.core/src/net/sf/eclipsecs/core/transformer/XmlProfileWriter.java
+++ b/net.sf.eclipsecs.core/src/net/sf/eclipsecs/core/transformer/XmlProfileWriter.java
@@ -1,0 +1,124 @@
+//============================================================================
+//
+// Copyright (C) 2002-2016  David Schneider, Lars Ködderitzsch
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+//
+//============================================================================
+
+package net.sf.eclipsecs.core.transformer;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.io.StringWriter;
+import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+import java.util.Map;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.OutputKeys;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+
+import net.sf.eclipsecs.core.util.CheckstyleLog;
+
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+
+/**
+ * Utility class to write eclipse formatter/cleanup profile XML files.
+ *
+ * @author Alexandros Karypidis
+ *
+ */
+public final class XmlProfileWriter {
+  private static final String XML_NODE_ROOT = "profiles";
+  private static final String XML_NODE_PROFILE = "profile";
+  private static final String XML_NODE_SETTING = "setting";
+
+  private static final String XML_ATTRIBUTE_VERSION = "version";
+  private static final String XML_ATTRIBUTE_ID = "id";
+  private static final String XML_ATTRIBUTE_NAME = "name";
+  private static final String XML_ATTRIBUTE_PROFILE_KIND = "kind";
+  private static final String XML_ATTRIBUTE_VALUE = "value";
+
+  private static final String CLEANUP_PROFILE_VERSION = "2";
+  private static final String CLEANUP_PROFILE_KIND = "CleanUpProfile";
+
+  private static final String FORMATTER_PROFILE_VERSION = "10";
+  private static final String FORMATTER_PROFILE_KIND = "CodeFormatterProfile";
+
+  private XmlProfileWriter() {
+    // no code
+  }
+
+  public static InputStream writeCleanupProfileToStream(String name,
+          Map<String, String> settings) throws TransformerException, ParserConfigurationException {
+    return writeProfileToStream(name, CLEANUP_PROFILE_VERSION, CLEANUP_PROFILE_KIND, settings);
+  }
+
+  public static InputStream writeFormatterProfileToStream(String name,
+          Map<String, String> settings) throws TransformerException, ParserConfigurationException {
+    return writeProfileToStream(name, FORMATTER_PROFILE_VERSION, FORMATTER_PROFILE_KIND, settings);
+  }
+
+  private static InputStream writeProfileToStream(String name, String profileVersion,
+          String profileKind, Map<String, String> settings) throws TransformerException,
+          ParserConfigurationException {
+    final DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+    final DocumentBuilder builder = factory.newDocumentBuilder();
+    final Document document = builder.newDocument();
+
+    final Element rootElement = document.createElement(XML_NODE_ROOT);
+    rootElement.setAttribute(XML_ATTRIBUTE_VERSION, profileVersion);
+
+    document.appendChild(rootElement);
+
+    final Element profileElement = document.createElement(XML_NODE_PROFILE);
+    profileElement.setAttribute(XML_ATTRIBUTE_NAME, name);
+    profileElement.setAttribute(XML_ATTRIBUTE_VERSION, profileVersion);
+    profileElement.setAttribute(XML_ATTRIBUTE_PROFILE_KIND, profileKind);
+
+    final Iterator<String> keyIter = settings.keySet().iterator();
+
+    while (keyIter.hasNext()) {
+      final String key = keyIter.next();
+      final String value = settings.get(key);
+      if (value != null) {
+        final Element setting = document.createElement(XML_NODE_SETTING);
+        setting.setAttribute(XML_ATTRIBUTE_ID, key);
+        setting.setAttribute(XML_ATTRIBUTE_VALUE, value);
+        profileElement.appendChild(setting);
+      } else {
+        CheckstyleLog.log(null, String.format("Profile is missing value for [key=%s]", key));
+      }
+    }
+    rootElement.appendChild(profileElement);
+
+    final Transformer transformer = TransformerFactory.newInstance().newTransformer();
+    transformer.setOutputProperty(OutputKeys.METHOD, "xml");
+    transformer.setOutputProperty(OutputKeys.ENCODING, StandardCharsets.UTF_8.name());
+    transformer.setOutputProperty(OutputKeys.INDENT, "yes");
+    final StringWriter writer = new StringWriter();
+    transformer.transform(new DOMSource(document), new StreamResult(writer));
+    return new ByteArrayInputStream(writer.toString().getBytes());
+  }
+
+}

